### PR TITLE
Streams and configuration implementation

### PR DESCRIPTION
**Pull request by valerio8787**

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -7,6 +7,7 @@
         "league/pipeline": "^0.3.0",
         "spryker/console": "^4.0.0",
         "spryker/development": "^3.0.0",
+        "spryker/log": "^3.0.0",
```
Comment (PhilinTv) [inline composer.json]: php 7.1

File: src/SprykerMiddleware/Shared/Process/Transfer/process.transfer.xml [hunk 2]
```diff
@@ -5,6 +5,8 @@
 
     <transfer name="ProcessSettings">
         <property name="name" type="string"/>
+        <property name="inputPath" type="string"/>
+        <property name="outputPath" type="string"/>
```
Comment (PhilinTv) [inline src/SprykerMiddleware/Shared/Process/Transfer/process.transfer.xml]: `iteratorSettings` is really a bad name for a singular parameter. We use plural only for arrays of smth. 
You could use smth like 'iteratorConfiguration' or 'iteratorConfig'

File: src/SprykerMiddleware/Zed/Process/Business/Exception/StagePluginsNotConfiguredException.php [hunk 3]
```diff
@@ -0,0 +1,38 @@
+<?php
+
+namespace SprykerMiddleware\Zed\Process\Business\Exception;
+
+use Exception;
+use Spryker\Shared\Config\Config;
+use Spryker\Shared\Kernel\Exception\Backtrace;
+use Spryker\Shared\Kernel\KernelConstants;
+
+class StagePluginsNotConfiguredException extends Exception
```
Comment (PhilinTv) [inline src/SprykerMiddleware/Zed/Process/Business/Exception/StagePluginsNotConfiguredException.php]: Where do you use it?

File: src/SprykerMiddleware/Zed/Process/Communication/Plugin/Handler/StdErrStreamHandlerPlugin.php [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    }
+
+    /**
+     * @param array $records
+     *
+     * @return mixed
+     */
+    public function handleBatch(array $records)
+    {
+        return $this->getHandler()->handleBatch($records);
```
Comment (PhilinTv) [inline src/SprykerMiddleware/Zed/Process/Communication/Plugin/Handler/StdErrStreamHandlerPlugin.php]: This returns void.

File: src/SprykerMiddleware/Zed/Process/Communication/ProcessCommunicationFactory.php [hunk 5]
```diff
@@ -1,11 +1,98 @@
...
+     */
+    public function getMiddlewareLogProcessors()
+    {
+        return $this->getProvidedDependency(ProcessDependencyProvider::MIDDLEWARE_LOG_PROCESSORS);
+    }
+
+    /**
+     * @return \SprykerMiddleware\Zed\Process\Dependency\Plugin\Configuration\ProcessConfigurationPluginInterface[]
+     */
+    public function getDefaultProcessesPlugins()
```
Comment (PhilinTv) [inline src/SprykerMiddleware/Zed/Process/Communication/ProcessCommunicationFactory.php:77]: Why it's default?

File: src/SprykerMiddleware/Zed/Process/Communication/Plugin/Configuration/DefaultConfigurationProfilePlugin.php [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+namespace SprykerMiddleware\Zed\Process\Communication\Plugin\Configuration;
+
+use Spryker\Zed\Kernel\Communication\AbstractPlugin;
+use SprykerMiddleware\Zed\Process\Dependency\Plugin\Configuration\ConfigurationProfilePluginInterface;
+
+/**
+ * @method \SprykerMiddleware\Zed\Process\Business\ProcessFacadeInterface getFacade()
+ * @method \SprykerMiddleware\Zed\Process\Communication\ProcessCommunicationFactory getFactory()
+ */
+class DefaultConfigurationProfilePlugin extends AbstractPlugin implements ConfigurationProfilePluginInterface
```
Comment (PhilinTv) [inline src/SprykerMiddleware/Zed/Process/Communication/Plugin/Configuration/DefaultConfigurationProfilePlugin.php]: Why do you have a plugin to return other plugins?

File: src/SprykerMiddleware/Zed/Process/Business/Log/LoggerFactory.php [hunk 7]
```diff
@@ -0,0 +1,28 @@
+<?php
+
+namespace SprykerMiddleware\Zed\Process\Business\Log;
+
+use Spryker\Shared\Log\Config\LoggerConfigInterface;
+use Spryker\Shared\Log\LoggerFactory as SprykerLoggerFactory;
+
+class LoggerFactory
```
Comment (PhilinTv) [inline src/SprykerMiddleware/Zed/Process/Business/Log/LoggerFactory.php]: No other factories in business logic (BL). You are injecting another namespace deep in BL.
Dependency provider is there exactly for this reason.

File: src/SprykerMiddleware/Zed/Process/Business/Log/LoggerTrait.php [hunk 8]
```diff
@@ -0,0 +1,26 @@
+<?php
+
+namespace SprykerMiddleware\Zed\Process\Business\Log;
+
+use Spryker\Shared\Log\Config\LoggerConfigInterface;
+
+trait LoggerTrait
```
Comment (PhilinTv) [inline src/SprykerMiddleware/Zed/Process/Business/Log/LoggerTrait.php]: A hard copy of the logger trait does not resolve the issue.
Or we will find another solution or just use a logger from Log module. But no overhead.

File: src/SprykerMiddleware/Zed/Process/Business/Mapper/Map/MapInterface.php [hunk 9]
```diff
@@ -6,6 +6,10 @@
 
 interface MapInterface
 {
+    const ITEM_MAP = 'itemMap';
```
Comment (PhilinTv) [inline src/SprykerMiddleware/Zed/Process/Business/Mapper/Map/MapInterface.php]: How and where do you use them?

File: src/SprykerMiddleware/Zed/Process/Business/Pipeline/Pipeline.php [hunk 10]
```diff
@@ -11,23 +13,21 @@ class Pipeline implements PipelineInterface
...
      */
     protected $processor;
 
     /**
-     * @param \League\Pipeline\ProcessorInterface $processor
+     * @param \SprykerMiddleware\Zed\Process\Business\Pipeline\Processor\PipelineProcessorInterface $processor
      * @param \SprykerMiddleware\Zed\Process\Business\Pipeline\Stage\StageInterface[] $stages
      */
-    public function __construct(ProcessorInterface $processor, array $stages)
+    public function __construct(PipelineProcessorInterface $processor, array $stages)
```
Comment (PhilinTv) [inline src/SprykerMiddleware/Zed/Process/Business/Pipeline/Pipeline.php]: PipelineProcessorInterface should be reflected in var names (eg processor => pipelineProcessor)

File: src/SprykerMiddleware/Zed/Process/Business/Pipeline/Processor/FingersCrossedProcessor.php [hunk 11]
```diff
@@ -0,0 +1,26 @@
+<?php
+
+namespace SprykerMiddleware\Zed\Process\Business\Pipeline\Processor;
+
+use SprykerMiddleware\Shared\Process\Stream\ReadStreamInterface;
+use SprykerMiddleware\Shared\Process\Stream\WriteStreamInterface;
+
+class FingersCrossedProcessor implements PipelineProcessorInterface
```
Comment (PhilinTv) [inline src/SprykerMiddleware/Zed/Process/Business/Pipeline/Processor/FingersCrossedProcessor.php:13]: Do we copy the league/pipeline?

File: src/SprykerMiddleware/Service/Process/Model/StreamServiceInterface.php [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+namespace SprykerMiddleware\Service\Process\Model;
+
+use SprykerMiddleware\Shared\Process\Stream\ReadStreamInterface;
+use SprykerMiddleware\Shared\Process\Stream\WriteStreamInterface;
+
+interface StreamServiceInterface
+{
+    /**
+     * Specification:
+     * - Read data from given stream.
```
Comment (lazystar) [inline src/SprykerMiddleware/Service/Process/Model/StreamServiceInterface.php]: Where do we have the rules for the specs? In this case here it is just bloat and does not add information, imho.

**Review by PhilinTv — CHANGES_REQUESTED** [review]

File: src/SprykerMiddleware/Zed/Process/Business/Iterator/NullIterator.php [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+        return $this->inStream;
+    }
+
+    /**
+     * @inheritdoc
+     * @throws \Exception
+     */
+    public function key()
+    {
+        throw  new Exception('not supported');
```
Comment (lazystar) [inline src/SprykerMiddleware/Zed/Process/Business/Iterator/NullIterator.php]: Maybe a little more info or even a specific exception?

File: src/SprykerMiddleware/Zed/Process/Business/Stream/JsonStream.php [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+        $this->position = 0;
+
+        return json_decode($data, true);
+    }
+
+    /**
+     * @param int $offset
+     * @param int $whence
+     *
+     * @return bool|int
```
Comment (lazystar) [inline src/SprykerMiddleware/Zed/Process/Business/Stream/JsonStream.php]: When is it bool?

File: src/SprykerMiddleware/Zed/Process/Business/Stream/JsonStream.php [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+     */
+    protected function getNewPosition($offset, $whence)
```
Comment (lazystar) [inline src/SprykerMiddleware/Zed/Process/Business/Stream/JsonStream.php]: We do not use primitive typehints?

File: src/SprykerMiddleware/Zed/Process/Communication/Plugin/Processor/IntrospectionProcessorPlugin.php [hunk 16]
```diff
@@ -0,0 +1,29 @@
+<?php
+
+/**
+ * Copyright © 2016-present Spryker Systems GmbH. All rights reserved.
+ * Use of this software requires acceptance of the Evaluation License Agreement. See LICENSE file.
+ */
```
Comment (lazystar) [inline src/SprykerMiddleware/Zed/Process/Communication/Plugin/Processor/IntrospectionProcessorPlugin.php:6]: Sometimes we have this header, sometimes not.

Comment (valerio8787) [reply src/SprykerMiddleware/Zed/Process/Communication/Plugin/Processor/IntrospectionProcessorPlugin.php:6]: I added header to all files

File: src/SprykerMiddleware/Zed/Process/Business/Pipeline/Processor/FingersCrossedProcessor.php (hunk 11 above)
Comment (valerio8787) [reply src/SprykerMiddleware/Zed/Process/Business/Pipeline/Processor/FingersCrossedProcessor.php:13]: I got read from league/pipeline

**Pull request merged**